ipm: imx: send firmware ready reply

iuliana-prodan opened the pull request:
Send a fw ready reply message as soon as possible.
This is usually used on host side which is waiting for this message in order to establish the communication with the remote processor - see `imx_dsp_rproc` driver from Linux.
    
This can be enabled by `IPM_IMX_FW_READY_REPLY` config, which is by default N.
    
Set `CONFIG_IPM_IMX_FW_READY_REPLY` as Y for `openamp_rsc_table` sample, running on nxp_adsp_imx8m.

iuliana-prodan committed:
ipm: imx: send firmware ready reply to check for fw boot completion

Send a fw ready reply message as soon as possible.
This is usually used on host side which is waiting
for this message in order to establish the
communication with the remote processor - see
imx_dsp_rproc driver from Linux.

This can be enabled by IPM_IMX_FW_READY_REPLY config,
which is by default N.

Set CONFIG_IPM_IMX_FW_READY_REPLY as Y for
openamp_rsc_table sample, running on nxp_adsp_imx8m.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

carlocaione commented on drivers/ipm/Kconfig.imx:
```diff
@@ -64,3 +64,10 @@ config IPM_IMX_MAX_ID_VAL
+
+config IPM_IMX_FW_READY_REPLY
+	bool "FW_READY reply after boot"
+	default n
```
```suggestion
```

iuliana-prodan replied:
Why should I remove this?
I don't want this to be enabled for all platforms that are using ipm_imx.

Is it better to use:
`depends on SOC_FAMILY_NXP_ADSP` ?
(https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/xtensa/nxp_adsp/Kconfig#L6 - it refers to the DSP from i.MX targets)

carlocaione replied:
`n` is default already. No need to specify it.

carlocaione commented on drivers/ipm/Kconfig.imx:
```diff
@@ -64,3 +64,10 @@ config IPM_IMX_MAX_ID_VAL
+	default n
+	depends on IPM_IMX || IPM_IMX_REV2
+	help
+	  Send FW_READY reply after boot
```
after boot? Then why this is in the IPM driver?

iuliana-prodan replied:
I'll update the message here - after messaging unit initialization, because I need the MU to be up in order to trigger the interrupt.

carlocaione reviewed (changes requested):
While this PR is not problematic per se my impression is that we are kinda cheating here. Wouldn't much better just to introduce a new REMOTEPROC API class / subsystem in Zephyr and use that instead?

iuliana-prodan commented:
> While this PR is not problematic per se my impression is that we are kinda cheating here. Wouldn't much better just to introduce a new REMOTEPROC API class / subsystem in Zephyr and use that instead?

This is not a standard - to send a fw ready reply after the remote core is up. This is just for imx use-case. 

Other remoteproc Linux drivers (i.e. omap_remoteproc) are sending an echo request and this message will wait in the mailbox fifo until the remote core is up. 

Others are doing nothing.

Therefore, there is no "one way fits all" for host-remote handshake communication.
In remoteproc there's no confirmation for remote core booted.

carlocaione commented:
> This is not a standard - to send a fw ready reply after the remote core is up. This is just for imx use-case.
> 
> Others remoteproc Linux drivers (e.g. omap_remoteproc) are sending an echo request and this message will wait in the mailbox fifo until the remote core is up.
> 
> Others are doing nothing.
> 
> Therefore, there is no "one way fits all" for host-remote handshake communication. In remoteproc there's no confirmation for remote core booted.

What about the `notify` hook here https://github.com/OpenAMP/open-amp/blob/5891cb4b2e813d1c9d10def998952b062304e43f/lib/include/openamp/remoteproc.h#L413 ?

I'm not going to NACK this PR, this is fine to me. I'm just trying to understand if there is a generic way to do this.

arnopo commented:
Agree with @carlocaione,

The handshake should be performed at the application level. It probably relies on a mailbox, but

Today, similar mechanism is done in Linux but from the main to the remote processor. When the virtio device status has been updated.  A kick (notify) is sent by Linux rpmsg virtio driver to the coprocessor. The Linux remoteproc is in charge of sending this kick through the mailbox.  Associated code is available here: **https://elixir.bootlin.com/linux/latest/source/drivers/rpmsg/virtio_rpmsg_bus.c#L995**

In this PR the inverse is requested.
So your Linux kernel driver should probably have to block on start waiting  a mailbox kick from the coprocessor before initializing the rpmsg.
The coprocessor application (https://elixir.bootlin.com/zephyr/latest/C/ident/rpmsg_mng_task) would be in charge of sending a notification when it is ready to communicate. And then wait the resource table virtio device status update.

Using a config is nice for compatibility with other hardware

iuliana-prodan commented:
> Agree with @carlocaione,
> 
> The handshake should be performed at the application level. It probably relies on a mailbox, but
> 
> Today, similar mechanism is done in Linux but from the main to the remote processor. When the virtio device status has been updated. A kick (notify) is sent by Linux rpmsg virtio driver to the coprocessor. The Linux remoteproc is in charge of sending this kick through the mailbox. Associated code is available here: **https://elixir.bootlin.com/linux/latest/source/drivers/rpmsg/virtio_rpmsg_bus.c#L995**
> 
The i.MX DSP remoteproc  (imx_dsp_rproc) driver is implementing the generic kick callback - https://elixir.bootlin.com/linux/latest/source/drivers/remoteproc/imx_dsp_rproc.c and is called after starting the remote core - see https://elixir.bootlin.com/linux/latest/source/drivers/remoteproc/imx_dsp_rproc.c#L316.
But, this will return error, because the FW_READY flag was not set (https://elixir.bootlin.com/linux/latest/source/drivers/remoteproc/imx_dsp_rproc.c#L291).
This is set on rxdb callback (https://elixir.bootlin.com/linux/latest/source/drivers/remoteproc/imx_dsp_rproc.c#L496) which uses general purpose interrupts.

> In this PR the inverse is requested. So your Linux kernel driver should probably have to block on start waiting a mailbox kick from the coprocessor before initializing the rpmsg. 

It does, as mentioned above (https://elixir.bootlin.com/linux/latest/source/drivers/remoteproc/imx_dsp_rproc.c#L344)

> The coprocessor application (https://elixir.bootlin.com/zephyr/latest/C/ident/rpmsg_mng_task) would be in charge of sending a notification when it is ready to communicate. And then wait the resource table virtio device status update.
> 

So, @arnopo you're saying I should update the [openamp_rsc_table](https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/subsys/ipc/openamp_rsc_table/src/main_remote.c) sample?
This already sends a notify (using [ipm_send](https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/subsys/ipc/openamp_rsc_table/src/main_remote.c#L140)), but is using TX interrupts, while in Linux, I'm waiting for GP interrupts.
So I cannot use the IPM API.

Now, in `ipm_imx`, I'm calling directly from NXP_HAL the `MU_TriggerInterrupts()` used for GP interrupts.

> Using a config is nice for compatibility with other hardwares

iuliana-prodan commented:
> > Agree with @carlocaione,
> > The handshake should be performed at the application level. It probably relies on a mailbox, but
> > Today, similar mechanism is done in Linux but from the main to the remote processor. When the virtio device status has been updated. A kick (notify) is sent by Linux rpmsg virtio driver to the coprocessor. The Linux remoteproc is in charge of sending this kick through the mailbox. Associated code is available here: **https://elixir.bootlin.com/linux/latest/source/drivers/rpmsg/virtio_rpmsg_bus.c#L995**
> 
> The i.MX DSP remoteproc (imx_dsp_rproc) driver is implementing the generic kick callback - https://elixir.bootlin.com/linux/latest/source/drivers/remoteproc/imx_dsp_rproc.c and is called after starting the remote core - see https://elixir.bootlin.com/linux/latest/source/drivers/remoteproc/imx_dsp_rproc.c#L316. But, this will return error, because the FW_READY flag was not set (https://elixir.bootlin.com/linux/latest/source/drivers/remoteproc/imx_dsp_rproc.c#L291). This is set on rxdb callback (https://elixir.bootlin.com/linux/latest/source/drivers/remoteproc/imx_dsp_rproc.c#L496) which uses general purpose interrupts.
> 
> > In this PR the inverse is requested. So your Linux kernel driver should probably have to block on start waiting a mailbox kick from the coprocessor before initializing the rpmsg.
> 
> It does, as mentioned above (https://elixir.bootlin.com/linux/latest/source/drivers/remoteproc/imx_dsp_rproc.c#L344)
> 
> > The coprocessor application (https://elixir.bootlin.com/zephyr/latest/C/ident/rpmsg_mng_task) would be in charge of sending a notification when it is ready to communicate. And then wait the resource table virtio device status update.
> 
> So, @arnopo you're saying I should update the [openamp_rsc_table](https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/subsys/ipc/openamp_rsc_table/src/main_remote.c) sample? This already sends a notify (using [ipm_send](https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/subsys/ipc/openamp_rsc_table/src/main_remote.c#L140)), but is using TX interrupts, while in Linux, I'm waiting for GP interrupts. So I cannot use the IPM API.

I thought a little bit more and I can expand the IPM API with an `ipm_notify()`.
Give me a couple of days and I'll send a new version to this.

Thanks!

PS: If I won't be able to send it by Friday, I'll do it next week since I'll be OOO :)

carlocaione commented:
> I thought a little bit more and I can expand the IPM API with an `ipm_notify()`.

Why? Please don't.

A `notify` hook has nothing to do with the IPM device class. The IPM API is supposed to support an Inter-Processor communication hardware, a `notify` hook is something related to your transport protocol, not the hardware.

iuliana-prodan commented:
> > I thought a little bit more and I can expand the IPM API with an `ipm_notify()`.
> 
> Why? Please don't.
> 
> A `notify` hook has nothing to do with the IPM device class. The IPM API is supposed to support an Inter-Processor communication hardware, a `notify` hook is something related to your transport protocol, not the hardware.

But then how can I send (actually just set the interrupts I need) the fw reply? This is platform specific?
I add the `ipm_notify` and this will be implemented where needed - like, for now, only in `ipm_imx`.

As mention, I cannot use the `ipm_send` since is using TX interrupts, rather than GP interrupts that are expected in Linux.

carlocaione commented:
> But then how can I send (actually just set the interrupts I need) the fw reply? This is platform specific. I add the `ipm_notify` and this will be implemented where needed - like, for now, only in `ipm_imx`.

You realize that we cannot arbitrary extend every API device class because some random platform needs something on top right? This being platform specific is exactly the reason this hook makes no sense in the general API class.

You extend an API when the new API is going to support a wide variety of hardware and platforms and, to me, it looks like this is not the case.

The IPM class is about hardware, how do you use this hardware is not the driver concern really. You want something on top of that, an application, a subsystem, a framework, etc... that is using that driver to do something with it (notify, build packets, etc...).

arnopo commented:
> > > I thought a little bit more and I can expand the IPM API with an `ipm_notify()`.
> > 
> > 
> > Why? Please don't.
> > A `notify` hook has nothing to do with the IPM device class. The IPM API is supposed to support an Inter-Processor communication hardware, a `notify` hook is something related to your transport protocol, not the hardware.
> 
> But then how can I send (actually just set the interrupts I need) the fw reply? This is platform specific. I add the `ipm_notify` and this will be implemented where needed - like, for now, only in `ipm_imx`.
> 
> As mention, I cannot use the `ipm_send` since is using TX interrupts, rather than GP interrupts that are expected in Linux.

What about considering the GP interrupt as a specific channel of the `ipm_imx` (or a new instance of the ipm)?
you could implement specific action in your `ipm_imx` based on the channel ID.

Then in application you could define a config that is set to the channel ID value. 
if the CONFIG is set the application calls `ipm_send(ipm_handle, 0, CONFIG_OPENAMP_RSC_TABLE_KICK_ID, NULL, 0);` that trigs the GP interrupt.

carlocaione commented on drivers/ipm/Kconfig:
```diff
@@ -55,6 +55,12 @@ config XLNX_IPI
 	  Inter Processor Interrupt driver for AMD-Xilinx
 	  platforms such as ZynqMP Ultrascale+.
 
+config IPM_SEND_FW_READY_REPLY
+	hex "Send a FW_READY reply message"
+	default 0xFFFF
+	help
+	  Send FW_READY reply to check for FW boot completion
+
```
If you want you can move this into `Kconfig.imx`, that would be acceptable.

iuliana-prodan replied:
I added this new version as @arnopo suggested -  just for all of us to review it :)
But, I prefer the first version of this PR - adding a config just for IPM_IMX.
Since the fw_ready is waited only for IMX we should fix this limitation only for imx.

@arnopo what do you think?

iuliana-prodan replied:
I reverted to my previous version (with small changes on commit message and CONFIG description).

I added the config only for _i.MX family_.
I enable it only for the board I need (right now this is only for the `openamp_rsc_table` sample compiled for `nxp_adsp_imx8m`)

arnopo reviewed (dismissed):
I am still not convinced that the solution is reliable in the long term (for instance, for a Zephyr project that does not use ipm driver). However, the patch should be able to respond in the short term.

dbaluta commented:
> I am still not convinced that the solution is reliable in the long term (for instance, for a Zephyr project that does not use ipm driver). However, the patch should be able to respond in the short term.

I'm OK with this for now, as it seems there is no general agreement with Linux folks. 

On the long run I think we can make use of `struct resource_table` https://github.com/OpenAMP/open-amp/blob/e63d07d10577b06c5b3fc8e1c20ede74779d3132/lib/include/openamp/remoteproc.h#L4

We could define a new `fw_resource_type` for example `RSC_PLATFORM`. 

All application will need to fill this struct with information about expected behavior. Like, kernel should wait for firmware to boot (or not).

It looks like the Linux kernel infrastructure is already in place. What we need to figure out is how to define such structure with open-amp and zephyr.

arnopo commented:
> > I am still not convinced that the solution is reliable in the long term (for instance, for a Zephyr project that does not use ipm driver). However, the patch should be able to respond in the short term.
> 
> I'm OK with this for now, as it seems there is no general agreement with Linux folks.
> 
> On the long run I think we can make use of `struct resource_table` https://github.com/OpenAMP/open-amp/blob/e63d07d10577b06c5b3fc8e1c20ede74779d3132/lib/include/openamp/remoteproc.h#L4
> 
> We could define a new `fw_resource_type` for example `RSC_PLATFORM`.
> 
> All application will need to fill this struct with information about expected behavior. Like, kernel should wait for firmware to boot (or not).
> 
> It looks like the Linux kernel infrastructure is already in place. What we need to figure out is how to define such structure with open-amp and zephyr.

Yes that a way to do it.  I don't know if that would cover all use cases (XIP, start by U-boot or secure context, Zephyr <->Zephyr IPC, ...).  This would also imply that the resource table is defined for all zephyr samples. But also, that we respect cache line to make the resource table R/W for both processors during the initialization phase.

For your information we started a work in openAMP community to integrate virtio drivers/devices for IPC and the virtio-mmio transport in place of the virtio remoteproc (so no resource table). We identify similar handshake need to ensure that the coprocessor is running before probing the Linux  virtio-mmio driver.

That's why synchronization using signaling seems to me more generic. But perhaps we will need different solutions to address all hardwares and use cases.

iuliana-prodan commented:
> > I am still not convinced that the solution is reliable in the long term (for instance, for a Zephyr project that does not use ipm driver). However, the patch should be able to respond in the short term.
> 
> I'm OK with this for now, as it seems there is no general agreement with Linux folks.
> 
> On the long run I think we can make use of `struct resource_table` https://github.com/OpenAMP/open-amp/blob/e63d07d10577b06c5b3fc8e1c20ede74779d3132/lib/include/openamp/remoteproc.h#L4
> 
> We could define a new `fw_resource_type` for example `RSC_PLATFORM`.
> 
> All application will need to fill this struct with information about expected behavior. Like, kernel should wait for firmware to boot (or not).
> 

The resource table, both in Linux and Zephyr, uses fw_rsc_vdev struct (meaning RSC_VDEV).
I cannot just add a new one and use that instead or both - this imply a serious refactorization - in Linux and openAMP.
Maybe, you mean to extend the fw_rsc_vdev struct. 

Probably, a solution will be to use the config space from fw_rsc_vdev struct, 
But this is nowhere  (Linux, Zephyr, OpenAMP) used - the [config_len](https://github.com/OpenAMP/open-amp/blob/e63d07d10577b06c5b3fc8e1c20ede74779d3132/lib/include/openamp/remoteproc.h#L299) is declared 0 and that's all.

> It looks like the Linux kernel infrastructure is already in place. What we need to figure out is how to define such structure with open-amp and zephyr.

Where is this? Can you please give me some links?

There is nothing in place, otherwise I would have used it.

dbaluta commented on drivers/ipm/ipm_imx.c:
```diff
@@ -321,6 +321,22 @@ static int imx_mu_init(const struct device *dev)
+	 * FW_READY reply.
+	 */
+	MU_Type *base = MU(config);
+
```
This looks like a false positive.

dleach02 commented:
@nashif, We appear to have a false-positive on the compliance. What should we do here?

https://github.com/zephyrproject-rtos/zephyr/pull/61709/files#diff-7a9de38974d6ca2f0d251849b385035241af755de445ddee41ab93fb33cd5d25R332

iuliana-prodan commented:
I've added a new version where I've fixed the (false positive) compliance check -https://github.com/zephyrproject-rtos/zephyr/pull/61709/files#diff-7a9de38974d6ca2f0d251849b385035241af755de445ddee41ab93fb33cd5d25R332

cc: @carlocaione @arnopo @nashif @dleach02

carlescufi commented:
@arnopo could you please take another look?

arnopo commented on drivers/ipm/ipm_imx.c:
```diff
@@ -321,6 +321,22 @@ static int imx_mu_init(const struct device *dev)
+	 * after starting the remote processor, the host is waiting for a
+	 * FW_READY reply.
+	 */
+	MU_Type * base = MU(config);
```
remove space between `*` and `base`

arnopo reviewed (commented):
only one minor comment else LGTM
